### PR TITLE
User is able to enter data in forms before session expiry flow could execute given connectivity is there

### LINKED ISSUE #958: User is able to enter data in forms before session expiry flow could execute given connectivity is there
Either user should be able to save data entered upon submit or delay should not be long enough to allow user to keep entering data.

### PR DESCRIPTION
**Pull request by syedowaisali90**
close #958

### CODE REVIEW

File: rdt-open-reader/src/main/java/io/ona/rdt/util/RDTJsonFormUtils.java [hunk 1]
```diff
@@ -239,6 +239,7 @@ public void startJsonForm(JSONObject form, Activity context, int requestCode) {
         intent.putExtra(JSON_FORM_PARAM_JSON, form.toString());
         intent.putExtra(PERFORM_FORM_TRANSLATION, true);
         context.startActivityForResult(intent, requestCode);
+        RDTApplication.getInstance().clearCurrActivityReference(context);
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/util/RDTJsonFormUtils.java:242]: Why is this cleared here?

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java [hunk 2]
```diff
@@ -267,9 +268,9 @@ protected Void doInBackground(Void... voids) {
             boolean isUserAuthorized = syncUtils.verifyAuthorization();
 
             Activity currentActivity = RDTApplication.getInstance().getCurrentActivity();
-            boolean isFormActivity = currentActivity instanceof RDTJsonFormActivity;
+            boolean isFormSourceActivity = (currentActivity instanceof PatientProfileActivity) || (currentActivity instanceof PatientRegisterActivity);
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:271]: Why are we checking for patientprofileactivity?

**Comment by vincent-karuri** [comment]
@syedowaisali90 Just to confirm, what was causing the issue described https://github.com/onaio/rdt-standard/issues/958#issuecomment-831330093 here? Does this address it and how?

File: rdt-open-reader/src/main/java/io/ona/rdt/util/RDTJsonFormUtils.java (hunk 1 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/RDTJsonFormUtils.java:242]: now, form loading completed and we are starting the form activity so we don't need to persist source activity.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java (hunk 2 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:271]: https://github.com/onaio/rdt-standard/issues/958#issuecomment-831758094

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java [hunk 3]
```diff
@@ -37,16 +38,17 @@
...
         populatePatientDetails();
         addListeners();
         setUpTabs();
         initializeFormWidgetKeyToTextMap();
     }
 
     @Override
-    protected void onResume() {
-        super.onResume();
-        Utils.verifyUserAuthorization(this);
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java:49]: Why remove this?

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java [hunk 4]
```diff
@@ -267,9 +267,9 @@ protected Void doInBackground(Void... voids) {
             boolean isUserAuthorized = syncUtils.verifyAuthorization();
 
             Activity currentActivity = RDTApplication.getInstance().getCurrentActivity();
-            boolean isFormActivity = currentActivity instanceof RDTJsonFormActivity;
+            boolean isFormSourceActivity = currentActivity instanceof PatientProfileActivity;
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:270]: Change this to `boolean isRegisterActivity = currentActivity instanceof CovidJsonFormActivity;`

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java [hunk 5]
```diff
@@ -267,9 +267,9 @@ protected Void doInBackground(Void... voids) {
...
 
-            if (!isUserAuthorized && !isFormActivity) {
+            if (!isUserAuthorized && !isFormSourceActivity) {
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:272]: And this to `if (!isUserAuthorized && isRegisterActivity)`

Comment (vincent-karuri) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:272]: to only logout when on the register page

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java [hunk 6]
```diff
@@ -37,16 +38,17 @@
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+        RDTApplication.getInstance().setCurrentActivity(this);
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java:41]: Move this to the register activity.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java (hunk 4 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java]: fixed.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java (hunk 3 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java]: reverted.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java (hunk 5 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java]: fixed.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java (hunk 6 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java]: done.

Comment (vincent-karuri) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java]: Add unit tests to verify that if the current activity is patientregister, logout is not called.

Comment (vincent-karuri) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java]: Also add a test to verify that patientregister sets itself as the currentactivity onresume and clears it onpause.

Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientProfileActivity.java]: test cases added

File: rdt-open-reader/src/test/java/io/ona/rdt/robolectric/util/UtilsTest.java [hunk 7]
```diff
@@ -275,6 +279,23 @@ public void testExceptionIsCaughtFromUserAuthorizationTask() throws Exception {
         ReflectionHelpers.setField(userAuthorizationVerificationTask, "syncUtils", syncUtils);
         userAuthorizationVerificationTask.execute();
 
+        ReflectionHelpers.setStaticField(Utils.UserAuthorizationVerificationTask.class, "INSTANCE", null);
+        RDTApplication.getInstance().clearCurrActivityReference(patientRegisterActivity);
+    }
+
+    @Test
+    public void testLogoutShouldNotCallWhenCurrentActivityIsNotPatientRegisterActivity() throws Exception {
```
Comment (vincent-karuri) [inline rdt-open-reader/src/test/java/io/ona/rdt/robolectric/util/UtilsTest.java:287]: rename to `testUVerifyUserAuthorizationShouldNotLogoutUserIfNotOnRegistrationPage`

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/activity/CovidPatientRegisterActivityTest.java [hunk 8]
```diff
@@ -44,6 +45,19 @@ public void setUp() {
                 .get();
     }
 
+    @Test
+    public void testVerifyGlobalActivityShouldBeCovidPatientRegisterActivity() throws Exception {
```
Comment (vincent-karuri) [inline rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/activity/CovidPatientRegisterActivityTest.java:49]: rename to `testVerifyCurrentActivityShouldBeCovidPatientRegisterActivity`

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/activity/CovidPatientRegisterActivityTest.java [hunk 9]
```diff
@@ -44,6 +45,19 @@ public void setUp() {
...
+        RDTApplication.getInstance().clearCurrActivityReference(covidPatientRegisterActivity);
+        Whitebox.invokeMethod(covidPatientRegisterActivity, "onResume");
+        Assert.assertThat(RDTApplication.getInstance().getCurrentActivity(), CoreMatchers.instanceOf(CovidPatientRegisterActivity.class));
+    }
+
+    @Test
+    public void testVerifyGlobalActivityShouldBeNull() throws Exception {
```
Comment (vincent-karuri) [inline rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/activity/CovidPatientRegisterActivityTest.java:56]: rename to `testOnPauseShouldClearCurrentActivity`

File: rdt-open-reader/src/test/java/io/ona/rdt/robolectric/util/UtilsTest.java (hunk 7 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/test/java/io/ona/rdt/robolectric/util/UtilsTest.java]: done

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/activity/CovidPatientRegisterActivityTest.java (hunk 8 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/activity/CovidPatientRegisterActivityTest.java]: done

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/activity/CovidPatientRegisterActivityTest.java (hunk 9 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/activity/CovidPatientRegisterActivityTest.java]: done

**Review by vincent-karuri — APPROVED** [review]

**Pull request merged**